lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,293 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'avro/io'
+require 'json'
+
+module Avro
+  module IO
+    # JsonDecoder keeps track of the data and reads it using a given data type.
+    class JsonDecoder
+      # memo-ized map of Schema::PrimitiveSchema objects so we're not allocating an object for every validation
+      PRIMITIVE_SCHEMAS = Hash[Schema::PRIMITIVE_TYPES_SYM.map { |k| [k, Schema::PrimitiveSchema.new(k)] }]
+
+      attr_accessor :data
+
+      def io_reader(reader)
+        datum = reader.read
+        datum = '""' if datum.empty?
+        data = JSON.parse(datum, {:quirks_mode => true})
+        @data = data
+
+        self
+      end
+
+      def read_null
+        validate(:null, data)
+        data
+      end
+
+      def read_boolean
+        validate(:boolean, data)
+        data
+      end
+
+      def read_int
+        validate(:int, data)
+        Integer(data)
+      end
+
+      def read_long
+        validate(:long, data)
+        Integer(data)
+      end
+
+      def read_float
+        validate(:float, data)
+        Float(data)
+      end
+
+      def read_double
+        read_float
+      end
+
+      def read_bytes
+        validate(:bytes, data)
+        data.encode(Encoding::ISO_8859_1, Encoding::UTF_8)
```

Comment:
As per [Avro spec] (https://avro.apache.org/docs/1.7.7/spec.html#schema_record), `bytes and fixed fields are JSON strings, where Unicode code points 0-255 are mapped to unsigned 8-bit byte values 0-255.`. The reason why encoding is changed here is because 0-255 character set is same in both ISO_8859_1 and Unicde and ISO_8859_1 encodes using 8 bit. This matches the Avro spec requirements. This is very similar to how Java [encodes](https://github.com/apache/avro/blob/trunk/lang/java/avro/src/main/java/org/apache/avro/io/JsonEncoder.java#L219) and [decodes](https://github.com/apache/avro/blob/trunk/lang/java/avro/src/main/java/org/apache/avro/io/JsonDecoder.java#L264) JSON bytes and fixed fields.

IRB behavior is same as encoding behavior here - 
```
irb(main):001:0> "\\u0063"
=> "\\u0063"
irb(main):002:0> "\\u0063".encode(Encoding::ISO_8859_1, Encoding::UTF_8)
=> "\\u0063"
irb(main):003:0> "\u0063"
=> "c"
irb(main):004:0> "\u0063".encode(Encoding::ISO_8859_1, Encoding::UTF_8)
=> "c"
```

Can you help me understand what would you have expected differently?